localstack/services/secretsmanager/provider.py

```diff
@@ -68,143 +84,458 @@ def __init__(self):
+    ]
+    for version_no_stages in versions_no_stages:
+        del self.versions[version_no_stages]
+
+
+@patch(SecretsManagerBackend.rotate_secret)
+def backend_rotate_secret(
+    _,
+    self,
+    secret_id,
+    client_request_token=None,
+    rotation_lambda_arn=None,
+    rotation_rules=None,
+):
+    rotation_days = "AutomaticallyAfterDays"
+
+    if not self._is_valid_identifier(secret_id):
+        raise SecretNotFoundException()
+
+    if self.secrets[secret_id].is_deleted():
+        raise InvalidRequestException(
+            "An error occurred (InvalidRequestException) when calling the RotateSecret operation: You tried to \
+            perform the operation on a secret that's currently marked deleted."
+        )
+
+    if rotation_lambda_arn:
+        if len(rotation_lambda_arn) > 2048:
+            msg = "RotationLambdaARN " "must <= 2048 characters long."
+            raise InvalidParameterException(msg)
+
+    if rotation_rules:
+        if rotation_days in rotation_rules:
+            rotation_period = rotation_rules[rotation_days]
+            if rotation_period < 1 or rotation_period > 1000:
+                msg = "RotationRules.AutomaticallyAfterDays " "must be within 1-1000."
+                raise InvalidParameterException(msg)
+
+    secret = self.secrets[secret_id]
+
+    # The rotation function must end with the versions of the secret in
+    # one of two states:
+    #
+    #  - The AWSPENDING and AWSCURRENT staging labels are attached to the
+    #    same version of the secret, or
+    #  - The AWSPENDING staging label is not attached to any version of the secret.
+    #
+    # If the AWSPENDING staging label is present but not attached to the same
+    # version as AWSCURRENT then any later invocation of RotateSecret assumes
+    # that a previous rotation request is still in progress and returns an error.
+    try:
+        version = next(
+            version
+            for version in secret.versions.values()
+            if AWSPENDING in version["version_stages"]
+        )
+        if AWSCURRENT in version["version_stages"]:
+            msg = "Previous rotation request is still in progress."
+            raise InvalidRequestException(msg)
+
+    except StopIteration:
+        # Pending is not present in any version
+        pass
+
+    if client_request_token:
+        self._client_request_token_validator(client_request_token)
+        new_version_id = client_request_token
+    else:
+        new_version_id = str(uuid.uuid4())
+
+    # We add the new secret version as "pending". The previous version remains
+    # as "current" for now. Once we've passed the new secret through the lambda
+    # rotation function (if provided) we can then update the status to "current".
+    self._add_secret(
+        secret_id,
+        None,
+        description=secret.description,
+        tags=secret.tags,
+        version_id=new_version_id,
+        version_stages=[AWSPENDING],
+    )
```

Comment:
Secret version referenced by the rotation's ClientRequestToken is required to be bounded to AWSPENDING at the start of secret rotation https://github.com/aws-samples/aws-secrets-manager-rotation-lambdas/blob/252449551d58075f353444743e81a8c56d9f96db/SecretsManagerRotationTemplate/lambda_function.py#L52.
Removing these lines would require the rotation to always be preceded by the manual insertion of a AWSPENDING version: I don't see this being the intended implementation, but do let me know if I am mistaken.

I understand your concern with this insertion is it would not lead to failure when retrieving the value bounded to AWSPENDING during creation: https://github.com/aws-samples/aws-secrets-manager-rotation-lambdas/blob/252449551d58075f353444743e81a8c56d9f96db/SecretsManagerRotationTemplate/lambda_function.py#L94. Instead this will lead to at least boto3 failing. I do however see all other non base implementation of secret rotation manually raising the exception based on their internal logic, eg. https://github.com/aws-samples/aws-secrets-manager-rotation-lambdas/blob/252449551d58075f353444743e81a8c56d9f96db/SecretsManagerActiveDirectoryRotationSingleUser/lambda_function.py#L209, https://github.com/aws-samples/aws-secrets-manager-rotation-lambdas/blob/252449551d58075f353444743e81a8c56d9f96db/SecretsManagerRDSMySQLRotationMultiUser/lambda_function.py#L115, etc.

Reply:
@MEPalma I see what your saying that it looks like that example expects the Secret to exist, but in order for the same code to create a new value for the secret, `GetSecretValue` must throw `ResourceNotFoundException`: https://github.com/aws-samples/aws-secrets-manager-rotation-lambdas/blob/252449551d58075f353444743e81a8c56d9f96db/SecretsManagerRotationTemplate/lambda_function.py#L96

Unless that happens, which it doesn't look like it is, this does not close #5645.  The code I'm using is running in AWS and rotating secrets as I'd expect, so the issue isn't with the assumptions I'm making about `ResourceNotFoundException` being thrown.

>  I do however see all other non base implementation of secret rotation manually raising the exception based on their internal logic

All of these examples expect `ResourceNotFoundException` to be thrown, but none of them are "manually raising the [ResourceNotFoundException] based on their internal logic".  I think what you're referring to is the [`KeyError`](https://github.com/aws-samples/aws-secrets-manager-rotation-lambdas/blob/252449551d58075f353444743e81a8c56d9f96db/SecretsManagerActiveDirectoryRotationSingleUser/lambda_function.py#L369) they raise, but the [code isn't catching that error to generate a new value for the secret](https://github.com/aws-samples/aws-secrets-manager-rotation-lambdas/blob/252449551d58075f353444743e81a8c56d9f96db/SecretsManagerRDSMySQLRotationMultiUser/lambda_function.py#L117).

The `ResourceNotFoundException` is defined and thrown/raised by the AWS client, not the internal logic of the lambda.

Reply:
If I'm understanding these examples properly, it looks like AWS will create the secret but set the value to null/nil/none.  If you try to fetch the secret value before setting the secret value, it will throw a `ResourceNotFoundException`.

Reply:
@c1moore I think you'll find this fix does induce the return of a `ResourceNotFoundException` in that scenario.

This is true as it promotes the execution of [this statement](https://github.com/spulec/moto/blob/master/moto/secretsmanager/models.py#L282) which binds type `ResourceNotFoundException` to the error response [here](https://github.com/spulec/moto/blob/5e55daebb8411bc41a1782fc1e4fa0ebd944f990/moto/secretsmanager/exceptions.py#L27). I have double checked this and seems to be the case.

`test_rotate_secret_with_lambda_2` is a unit test carrying out two rotations in `tests/integration/secretsmanager/test_secretsmanager.py` using a rotation lambda similar to the one you brought up `tests/integration/secretsmanager/functions/lambda_rotate_secret.py` (in this pr).

Reply:
I'm probably missing something since Python isn't my main language, but how is that line triggered.  If I'm understanding it properly, we're setting the string value of that temporary secret [here](https://github.com/spulec/moto/blob/master/moto/secretsmanager/models.py#L519).

Regardless, don't let this comment thread hold up this PR, it sounds like this scenario has been tested and is doing what's expected, this is now just for my understanding.

Reply:
After I write that, I see that we're setting the string value to `None` in the provider.py, so how does provider.py and model.py interact?  It seems like model.py still has this bug/issue, but provider.py doesn't.